site/ja/guide/create_op.md

```diff
@@ -0,0 +1,1231 @@
+
+どのように取り組むのかを見るために、`int32` のテンソルを受け取り、最初以外のすべての要素が0であるコピーされたテンソルを出力するオペレーションを作ることを考えます。
+これを行うために、`zero_out.cc` と命名されたファイルを作成します。
+続いて、オペレーションのインターフェースを定義するための `REGISTER_OP` マクロ呼び出しを追加します。
+
+```c++
+#include "tensorflow/core/framework/op.h"
+#include "tensorflow/core/framework/shape_inference.h"
+
+using namespace tensorflow;
+
+REGISTER_OP("ZeroOut")
+    .Input("to_zero: int32")
+    .Output("zeroed: int32")
+    .SetShapeFn([](::tensorflow::shape_inference::InferenceContext* c) {
+      c->set_output(0, c->input(0));
+      return Status::OK();
+    });
+```
+
+この `ZeroOut` オペレーションは、入力として32bit整数のテンソル `to_zero` を受け取り、32bit整数のテンソル `zeroed` を出力します。
+このオペレーションは、出力テンソルが入力テンソルとおなじシェイプであることを保証するために、シェイプ関数を使っています。
+たとえば、入力テンソルのシェイプが [10, 20] であるならば、このシェイプ関数は出力のシェイプも [10, 20] であることを明示します。
+
+> 命名に関する注釈: オペレーションの名前はCamelCaseで、かつバイナリに登録されているすべてのオペレーションの中で唯一のものである必要があります。
+
+
+## オペレーションのカーネル実装
+
+インターフェースを定義したあとは、1つ以上のオペレーションの実装を提供する必要があります。
+これらのカーネルを作成するためには、`OpKernel` を継承したクラスを作成し、`Compute` メソッドをオーバーライドします。
+`Compute` メソッドは、`OpKernelContext*` 型である1つの `context` 引数を提供し、ここから入力や出力テンソルのような便利なものにアクセスできます。
+
+上記で作成したファイルにカーネルを追加します。
+カーネルはたとえば次のようなものになるかもしれません。
+
+```c++
+#include "tensorflow/core/framework/op_kernel.h"
+
+using namespace tensorflow;
+
+class ZeroOutOp : public OpKernel {
+ public:
+  explicit ZeroOutOp(OpKernelConstruction* context) : OpKernel(context) {}
+
+  void Compute(OpKernelContext* context) override {
+    // 入力テンソルを取得する。
+    const Tensor& input_tensor = context->input(0);
+    auto input = input_tensor.flat<int32>();
+
+    // 出力テンソルを作成する。
+    Tensor* output_tensor = NULL;
+    OP_REQUIRES_OK(context, context->allocate_output(0, input_tensor.shape(),
+                                                     &output_tensor));
+    auto output_flat = output_tensor->flat<int32>();
+
+    // 最初以外のすべての要素を0にする。
+    const int N = input.size();
+    for (int i = 1; i < N; i++) {
+      output_flat(i) = 0;
+    }
+
+    // 可能なら、最初の入力値は維持する。
+    if (N > 0) output_flat(0) = input(0);
+  }
+};
+```
+
+カーネルを実装したあと、TensorFlowのシステムに登録します。
+登録時には、このカーネルが動作するいろいろな制約を指定します。
+たとえば、CPU向けに作成した1つのカーネルと、GPU向けの別のカーネルがあるとしましょう。
+
+これを `ZeroOut` オペレーションで実現するためには、次を `zero_out.cc` に追加します。
+
+```c++
+REGISTER_KERNEL_BUILDER(Name("ZeroOut").Device(DEVICE_CPU), ZeroOutOp);
+```
+
+> 重要: OpKernelのインスタンスは、同時にアクセスされることがあります。`Compute` メソッドは、スレッドセーフにしなければなりません。クラスメンバへのアクセスはmutexでガードしてください。いっそのこと、クラスメンバ経由で状態を共有しないようにしてください！オペレーションの状態を追跡するためには、[`ResourceMgr`](https://www.tensorflow.org/code/tensorflow/core/framework/resource_mgr.h) を使用することを検討してください。
+
+
+### マルチスレッド化されたCPUカーネル
+
+マルチスレッド化されたCPUカーネルを書くためには、[`work_sharder.h`](https://www.tensorflow.org/code/tensorflow/core/util/work_sharder.h) にあるShard関数を利用できます。
+この関数は、スレッド間で計算のための関数を分割し、オペレーション内のスレッド制御のために使われます。（[`config.proto`](https://www.tensorflow.org/code/tensorflow/core/protobuf/config.proto) のintra_op_parallelism_threadsを見てください）
```

Comment:
[weak suggestion]
```suggestion
この関数は、オペレーション内でのスレッド実行のため使われるスレッド間で計算を分割します。（[`config.proto`](https://www.tensorflow.org/code/tensorflow/core/protobuf/config.proto) のintra_op_parallelism_threadsを見てください）
```

Reply:
Your suggestion is more straightforward 👍